Automatic file splitting duplicate audio

norihiro opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
When splitting file (#5371), audio packets are duplicated and saved to both previous and new files.
In the new file, audio timestamps will be offset by video timestamp so that video and audio has the same alignment among the split files in one recording.

Currently, the implementation relays on a video encoder that has DTS < PTS to duplicate audio packet. If DTS = PTS, the audio won't be duplicated. This behavior should be resolved before removing draft state.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
Audio glitch is the issue in the new WIP feature Automatic File Splitting (#5371).
Since the audio has different frame size than that for video, it is not easy for the boundary audio packet to be sent into just one file without audio glitch.

This change will resolve audio glitch when editing on NLEs. However, since the audio packets are duplicated, ffmpeg `concat` input might not be able to recover the original video. This function should be an optional.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
OBS Studio is built with FFmpeg 4.4.1 on Fedora 34. Encoder setting is libx264.
Output video files are tested with Kdenlive 21.08.1.
Kdenlive sometimes treats the length of the file one frame longer, which results audio glitch. User has to manually shrink the video length to have smooth switch to the new file.

It would be welcome if someone can test with other platforms, encoder settings, and NLEs.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [ ] The code has been tested.
- [ ] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

flaeri commented:
Hello again :)
As you said you wanted more testers on this, here is my report.
Checked out this PR, and rebased on latest master at the time (5f91bef8758fc228c2bad42c0e21833fd6b6e62c)

Windows 10 21H2
https://obsproject.com/logs/d1tomaYChKYzxSPa

In Davinci Resolve, I am seeing the same/similar, one frame audio dropout between the splits, as before this change.
I redid this test 3 times, confirming the output was the same.

The top green is the audio generated when the split files are placed back to back in the NLE.
The bottom green is the audio when using ffmpeg concat (perfect, cannot hear or see the split)
![image](https://user-images.githubusercontent.com/50419942/162788418-f585d013-3c84-4852-984a-ffbe161b94a5.png)

Example output from NLE (a little bit loud). Listen around 8-9 sec:
https://user-images.githubusercontent.com/50419942/162790061-15825d69-187d-42e8-94a1-4d0748d3359b.mov

Files OBS generated:
https://send.tresorit.com/a#4M_xuHED3qVgaPjyly02KQ

norihiro commented:
Implemented another idea; split only at the timing when both video and audio has the same timestamp.
The implementation relies on 60 fps video and 48 kHz audio setting. It requires more implementation for the case video-frame > audio-frame.
Another concern is that such condition might not occur frequently especially 59.94 fps and/or 44.1 kHz audio. Also note that some encoder could generate key frames randomly, which results much harder to meet the split condition. Maybe need to force encoder to output keyframe at the certain frame.

Picture below is a test results with 2 stitching points.
Added 432 Hz and 431 Hz tones to OBS Studio and recorded with the automatic split file feature. Then, stitched by Kdenlive.
First and second tracks are 432 Hz and 431 Hz tone, respectively. Third track is a stereo track  with 432 Hz and 431 Hz tones. The last track shows difference of OBS's recorded track and tone tracks. It shows no audio gap or glitch at all.
![a](https://user-images.githubusercontent.com/780600/168861370-972fdbf0-8d0e-4bb8-92d9-422d0623a206.png)

PatTheMav commented:
@norihiro What's the state of this PR? @flaeri's comment makes it seem as if the original implementation does not fix the issue and you came up with a different idea?

Is there a working implementation that fixes the issue? In that case I'd ask for the branch to be rebased and pushed and the PR undrafted so it can be reviewed.
